clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -47,32 +231,93 @@ public boolean isValid() {
         }
     }
 
-    public TransactionState(Time time) {
-        this.pidAndEpoch = new PidAndEpoch(NO_PRODUCER_ID, NO_PRODUCER_EPOCH);
-        this.sequenceNumbers = new HashMap<>();
-        this.time = time;
+    public synchronized FutureTransactionalResult initializeTransactions() {
+        ensureTransactional();
+        if (!transitionTo(State.INITIALIZING))
+            throw new IllegalStateException("Could not initialize transactions. Either transactions have already been " +
+                    "initialized or are being initialized.");
+        setPidAndEpoch(NO_PRODUCER_ID, NO_PRODUCER_EPOCH);
+        this.sequenceNumbers.clear();
+        if (transactionCoordinator == null)
+            pendingTransactionalRequests.add(findCoordinatorRequest(FindCoordinatorRequest.CoordinatorType.TRANSACTION, transactionalId, false));
+
+        TransactionalRequestResult result = new TransactionalRequestResult();
+        FutureTransactionalResult resultFuture = new FutureTransactionalResult(result);
+        if (!hasPid())
+            pendingTransactionalRequests.add(initPidRequest(false, result));
+        else
+            result.done();
+
+        return resultFuture;
     }
 
-    public boolean hasPid() {
-        return pidAndEpoch.isValid();
+    public synchronized void beginTransaction() {
+        ensureTransactional();
+        if (!transitionTo(State.IN_TRANSACTION))
+             throw new IllegalStateException("Producer isn't ready to begin a transaction.");
```

Comment:
Maybe the message should say that the producer already has an on-going transaction?